--- a/librocrand_common.jl	2020-11-08 15:33:32.901990031 +0100
+++ b/librocrand_common.jl	2020-11-08 17:01:20.335155791 +0100
@@ -16,7 +16,7 @@
 const ROCRAND_DEFAULT_MIN_WARPS_PER_EU = 1
 const rocrand_generator_base_type = Cvoid
 const rocrand_generator = Ptr{rocrand_generator_base_type}
-const half = Cint
+const half = Float16
 
 @cenum rocrand_status::UInt32 begin
     ROCRAND_STATUS_SUCCESS = 0
