--- a/librocfft_common.jl	2019-10-24 18:12:50.644135628 +0200
+++ b/librocfft_common.jl	2019-10-24 18:12:59.667664614 +0200
@@ -4,7 +4,6 @@
 const rocfft_version_major = 0
 const rocfft_version_minor = 9
 const rocfft_version_patch = 7
-const rocfft_version_tweak = ((((722 - rocm) - rel) - 2.9) - 6) - e3055e1
 const rocfft_plan_t = Cvoid
 const rocfft_plan = Ptr{rocfft_plan_t}
 const rocfft_plan_description_t = Cvoid
@@ -81,11 +81,3 @@
 
 
 const rocfft_layer_mode = rocfft_layer_mode_
-
-# Skipping MacroDefinition: ROCFFT_EXPORT __attribute__ ( ( visibility ( "default" ) ) )
-# Skipping MacroDefinition: ROCFFT_NO_EXPORT __attribute__ ( ( visibility ( "hidden" ) ) )
-# Skipping MacroDefinition: ROCFFT_DEPRECATED __attribute__ ( ( __deprecated__ ) )
-
-const ROCFFT_DEPRECATED_EXPORT = ROCFFT_EXPORT
-const ROCFFT_DEPRECATED_NO_EXPORT = ROCFFT_NO_EXPORT
-const DEFINE_NO_DEPRECATED = 0
